File: src/typescript/duck.ts

interface Duck {
	quack() : string;
	dance() : string;
}

class Wolf {
	quack() : string { return "QUACK QUACK WHOO"; }
	dance() : string { return " ¯\\_()_/¯ "; }
	eat(duck : Duck) : string { return " 😈 "; }
}

class Mallard implements Duck {
	quack() : string { return "quack quack" }
	dance() : string { return " _/¯ " }
}


const wolf : Wolf = new Wolf()
const theDuck : Duck = wolf // Not mandatory

const aDuck : Duck = new Mallard()

twoDucksAlone(aDuck, theDuck)

console.log(wolf.eat(aDuck))


function twoDucksAlone(d1:Duck, d2:Duck) {
	console.log(d1.quack())
	console.log(d2.quack())

	console.log(d1.dance())
	console.log(d2.dance())
}

File: src/haskell/duck.hs
{- A typeclass that Wolf and Mallard will have to explicitly instantiate -}
class Duck a where
    quake :: a -> String
    dance :: a -> String

data Wolf = Wolf

eat :: (Duck t) => Wolf -> t -> String
eat _ _ = " 😈 "

{- In order to become a duck, Wolf has to use the Adapter pattern. In
- Haskell, typeclasses are Constraints and not types, so it's impossible to
- force the type of any value to be Duck (it's not a type). And a Wolf
- is not a Duck, so it has to disguise himself. -}
data DuckCosplay = DuckCosplay Wolf
instance Duck DuckCosplay where
    quake _ = "QUACK QUACK WHOO"
    dance _ = " ¯\\_()_/¯ "

data Mallard = Mallard
instance Duck Mallard where
    quake _ = "quack quack"
    dance _ = " _/¯ "

twoDucksAlone :: (Duck aDuck, Duck theDuck) => aDuck -> theDuck -> IO()
twoDucksAlone aDuck theDuck = do
    putStrLn (quake aDuck)
    putStrLn (quake theDuck)

    putStrLn (dance aDuck)
    putStrLn (dance theDuck)


main = do
    let theWolf = Wolf
    let theDuck = DuckCosplay Wolf
    let aDuck = Mallard

    twoDucksAlone aDuck theDuck

    putStrLn (theWolf `eat` aDuck)
